conversion.go

```diff
@@ -16,37 +16,47 @@ const (
 	u256_nWords = 256 / bits.UintSize // number of Words in 256-bit
 )
 
-// marshallBigintGeneric is a platform-independent implementation of MarshallBigInt
+// NewFromBig is a platform-independent implementation of MarshallBigInt
 func NewFromBig(b *big.Int) (*Int, bool) {
+	fixed := &Int{}
+	overflow := fixed.SetFromBig(b)
+	return fixed, overflow
+}
+
+// SetFromBig
+// ensure we have sufficient testing, esp for negative bigints
+func (fixed *Int) SetFromBig(b *big.Int) bool {
 	var overflow bool
 	z := b.Bits()
 	numWords := len(z)
-	if numWords == 0 {
-		return &Int{}, overflow
-	}
+
 	// If there's more than 64 bits, we can skip all higher words
 	// z consists of 64 or 32-bit words. So we only care about the last
 	// (or last two)
 	if numWords > u256_nWords {
-		z = z[:u256_nWords]
-		numWords = len(z)
+		numWords = u256_nWords
 		overflow = true
 	}
-	// Code below is for 64-bit platforms only (numWords: [1-4] )
-	fixed := &Int{}
-	fixed[0] = uint64(z[0])
-	if numWords > 1 {
-		fixed[1] = uint64(z[1])
-		if numWords > 2 {
-			fixed[2] = uint64(z[2])
-			if numWords > 3 {
-				fixed[3] = uint64(z[3])
-			}
+
+	var w [8]big.Word // TODO: The size should be 4 or 8, but compiler complains. Move 32-bit implementation to separate function.
+	copy(w[:numWords], z[:numWords])
```

Comment:
Is this copy needd?

Reply:
I wanted to have the final step in the switch statement "fixed size" so I'm copying "variable length" on top of zeroed array here. But this copy is problematic, because it has bounds checks for both `w` and `z`.

I plan to start from beginning and reiterate. Goal: no panic checks.

Also I'm adding a benchmark for `SetFromBig()` up front.